File: sdks/typescript/packages/mpm-download/src/__tests__/testHarness.ts
// @ts-ignore
import { BaseBearerTokenAuthenticationProvider } from "@microsoft/kiota-abstractions";
// @ts-ignore
import { FetchRequestAdapter, KiotaClientFactory } from "@microsoft/kiota-http-fetchlibrary";
import { Gs1BeluAccessTokenProvider } from "../auth/accessTokenProvider.js";
import { SubscriptionKeyMiddleware } from "../auth/subscriptionKeyMiddleware.js";
import { BearerRetryMiddleware } from "../auth/bearerRetryMiddleware.js";
import { RateLimitMiddleware } from "../auth/rateLimitMiddleware.js";
import { Gs1BeluDownloadClient } from "../gs1BeluDownloadClient.js";
import type { Gs1BeluCredentials } from "../credentials.js";
import { FakeFetch } from "./fakeFetch.js";

export const FAKE_BASE_URL = "https://fake-api.test/myproductmanager/download/v17";

export interface TestHarnessOptions {
  credentials?: Gs1BeluCredentials;
  skewMarginMs?: number;
  rateLimitPerWindow?: number;
  rateLimitWindowMs?: number;
}

/**
 * The test seam Testing Decisions §"The seam" describes: builds the exact same production auth +
 * middleware pipeline as `Gs1BeluDownloadClient`'s derived-config constructor form, but wired to
 * fake HTTP transports (API + token endpoint) instead of the real network. Consumers never see
 * this — it exists only so tests can assert on observable HTTP behavior through the real classes.
 */
export function createTestHarness(options: TestHarnessOptions = {}) {
  const credentials: Gs1BeluCredentials = options.credentials ?? {
    clientId: "fake-client-id",
    clientSecret: "fake-client-secret",
    subscriptionKey: "fake-subscription-key",
  };
  const apiTransport = new FakeFetch();
  const tokenTransport = new FakeFetch();

  const tokenProvider = new Gs1BeluAccessTokenProvider(
    credentials,
    "https://fake-token.test/oauth/token",
    "https://fake-api.test/",
    ["fake-api.test"],
    { fetchFn: tokenTransport.fetch, skewMarginMs: options.skewMarginMs },
  );
  const authProvider = new BaseBearerTokenAuthenticationProvider(tokenProvider);

  const middlewares = [
    new SubscriptionKeyMiddleware(credentials.subscriptionKey),
    new BearerRetryMiddleware(tokenProvider),
    new RateLimitMiddleware(options.rateLimitPerWindow ?? 1000, options.rateLimitWindowMs),
  ];
  const httpClient = KiotaClientFactory.create(apiTransport.fetch, middlewares);
  const adapter = new FetchRequestAdapter(authProvider, undefined, undefined, httpClient);
  adapter.baseUrl = FAKE_BASE_URL;

  const client = new Gs1BeluDownloadClient({ requestAdapter: adapter });

  return { client, apiTransport, tokenTransport, tokenProvider };
}

File: sdks/dotnet/Gs1Belu.MyProductManager.Download.Tests/TestHarness.cs
using System;
using System.Net.Http;
using Gs1Belu.MyProductManager.Download.Auth;
using Microsoft.Kiota.Abstractions.Authentication;
using Microsoft.Kiota.Http.HttpClientLibrary;

namespace Gs1Belu.MyProductManager.Download.Tests;

/// <summary>
/// The test seam Testing Decisions §"The seam" describes: builds the exact same production auth +
/// middleware pipeline as <see cref="Gs1BeluDownloadClient"/>'s public constructor, but wired to fake
/// HTTP transports (API + token endpoint) instead of the real network. Consumers never see this —
/// it exists only so tests can assert on observable HTTP behavior through the real classes.
/// </summary>
internal sealed class TestHarness
{
    public const string FakeBaseUrl = "https://fake-api.test/myproductmanager/download/v17";

    public FakeHttpMessageHandler ApiTransport { get; } = new();
    public FakeHttpMessageHandler TokenTransport { get; } = new();
    public Gs1BeluAccessTokenProvider TokenProvider { get; }
    public Gs1BeluDownloadClient Client { get; }

    public TestHarness(
        Gs1BeluCredentials? credentials = null,
        TimeSpan? skewMargin = null,
        int rateLimitPerWindow = 1000,
        TimeSpan? rateLimitWindow = null)
    {
        credentials ??= new Gs1BeluCredentials("fake-client-id", "fake-client-secret", "fake-subscription-key");

        TokenProvider = new Gs1BeluAccessTokenProvider(
            credentials,
            new Uri("https://fake-token.test/oauth/token"),
            audience: "https://fake-api.test/",
            allowedHostsValidator: new AllowedHostsValidator(new[] { "fake-api.test" }),
            tokenTransportHandler: TokenTransport,
            skewMargin: skewMargin);

        var authProvider = new BaseBearerTokenAuthenticationProvider(TokenProvider);
        var handlers = new DelegatingHandler[]
        {
            new SubscriptionKeyHandler(credentials.SubscriptionKey),
            new BearerRetryHandler(TokenProvider),
            new RateLimitHandler(rateLimitPerWindow, rateLimitWindow),
        };
        var chainHead = KiotaClientFactory.ChainHandlersCollectionAndGetFirstLink(ApiTransport, handlers)
            ?? throw new InvalidOperationException("Kiota did not return a handler chain.");
        var httpClient = new HttpClient(chainHead);
        var adapter = new HttpClientRequestAdapter(authProvider, httpClient: httpClient) { BaseUrl = FakeBaseUrl };

        Client = new Gs1BeluDownloadClient(adapter);
    }

    public static HttpResponseMessage JsonResponse(System.Net.HttpStatusCode statusCode, string json) => new(statusCode)
    {
        Content = new StringContent(json, System.Text.Encoding.UTF8, "application/json"),
    };
}
